### PR TITLE
Fix latest tslint errors

### PR DESCRIPTION
**Pull request by SaschaNaz**
Except this type of errors:

```ts
export function getNavigationTree(sourceFile: SourceFile): NavigationTree {
    curSourceFile = sourceFile;
    const result = convertToTree(rootNavigationBarNode(sourceFile));
    curSourceFile = undefined;
    return result;
}
let curSourceFile: SourceFile;
// Identifier 'curSourceFile' is never reassigned; use 'const' instead of 'let'.
```

```ts
let response: protocol.Response;
try {
    ({ response } = this.executeCommand(msg));
}
//  Identifier 'response' is never reassigned; use 'const' instead of 'let'.
```

I think they are tslint bugs rather than actual code problem. (https://github.com/palantir/tslint/issues/1898)

BTW I feel fairly painful to create a new constant for every for-loop, I hope there is a better way to keep high performance...

### CODE REVIEW

**Comment by SaschaNaz** [comment]
[A benchmark shows for-of loop is much slower on Chrome](http://jsben.ch/#/arpvl), does this mean we shouldn't use for-of?

**Comment by andy-ms** [comment]
`for-of` gets emitted as a `for` loop, since we use an ES5 target. (So maybe some of those manually-written `for` loops could be upgraded.)
Also, this would be a good time to delete our own `preferConstRule.ts` since tslint has that now.

**Comment by SaschaNaz** [comment]
Already done the upgrade for obvious cases but many of them are using indices so I couldn't do it for them.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -5118,7 +5119,8 @@ namespace ts {
         function getSignaturesOfSymbol(symbol: Symbol): Signature[] {
             if (!symbol) return emptyArray;
             const result: Signature[] = [];
-            for (let i = 0, len = symbol.declarations.length; i < len; i++) {
+            const len = symbol.declarations.length;
+            for (let i = 0; i < len; i++) {
```
Comment (mhegazy) [inline src/compiler/checker.ts]: switch to `for..of`?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -5019,7 +5019,8 @@ namespace ts {
                 // If this is a JSDoc construct signature, then skip the first parameter in the
                 // parameter list.  The first parameter represents the return type of the construct
                 // signature.
-                for (let i = isJSConstructSignature ? 1 : 0, n = declaration.parameters.length; i < n; i++) {
+                const n = declaration.parameters.length;
+                for (let i = isJSConstructSignature ? 1 : 0; i < n; i++) {
                     const param = declaration.parameters[i];
```
Comment (mhegazy) [inline src/compiler/checker.ts]: ` for (let i = isJSConstructSignature ? 1 : 0; i < declaration.parameters.length; i++) {`

File: src/compiler/checker.ts (hunk 1 above)
Comment (SaschaNaz) [inline src/compiler/checker.ts]: Inner lines of code uses `i` for some purpose. It seems I can still switch to for-of in this case but should I do it?

**Comment by mhegazy** [comment]
@SaschaNaz i have merged a change to move tslint back to `4.0.0-dev.3` (https://github.com/Microsoft/TypeScript/pull/13028). can you update it with your PR to `next` again.

**Comment by SaschaNaz** [comment]
It still won't pass because of some bugs, which should be fixed on tslint side.

**Comment by mhegazy** [comment]
> It still won't pass because of some bugs, which should be fixed on tslint side.

do you have a list of the issues?

**Comment by andy-ms** [comment]
@mhegazy See first post and palantir/tslint#1898. I actually fixed one of them yesterday. The other one is also a legitimate issue, but could be fixed on our end by moving the declaration to before the function.

**Comment by SaschaNaz** [comment]
@andy-ms I'm still seeing destructuring issue on Travis, should I do something to get the merged fix?

**Comment by andy-ms** [comment]
`tslint` hasn't published a new version yet, so there's not much to do but wait.

**Comment by vladima** [comment]
I'd prefer not to change our code to workaround bugs in the toolchain hence https://github.com/palantir/tslint/pull/1908

**Comment by mhegazy** [comment]
@nchen63 any news about https://github.com/palantir/tslint/pull/1908?

**Comment by nchen63** [comment]
it's merged and included in a new release, v4.2.0

**Comment by SaschaNaz** [comment]
It seems `tslint@next` is still giving `tslint@4.1.0-dev.0`.

**Comment by nchen63** [comment]
should be fixed now

**Pull request merged**